mirrord-layer/src/tcp/outgoing.rs

```diff
@@ -0,0 +1,308 @@
+#[derive(Debug, Default)]
+pub(crate) struct TcpOutgoingHandler {
+    /// Holds the channels used to send daemon messages to the interceptor socket, for the case
+    /// where (agent) received data from the remote host, and sent it to (layer), to finally be
+    /// passed all the way back to the user.
+    mirrors: HashMap<ConnectionId, ConnectionMirror>,
+
+    /// Holds the connection requests from the `connect` hook. It's main use is to reply back with
+    /// the `SocketAddr` of the socket that'll be used to intercept the user's socket operations.
+    connect_queue: ResponseDeque<MirrorAddress>,
+}
+
+/// Wrapper type around `tokio::Sender`, used to send messages from the `agent` to our interceptor
+/// socket, where they'll be written back to the user's socket.
+///
+/// (agent) -> (layer) -> (user)
+#[derive(Debug)]
+pub(crate) struct ConnectionMirror(tokio::sync::mpsc::Sender<Vec<u8>>);
+
+impl Deref for ConnectionMirror {
+    type Target = tokio::sync::mpsc::Sender<Vec<u8>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for ConnectionMirror {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl TcpOutgoingHandler {
+    async fn interceptor_task(
+        connection_id: ConnectionId,
+        mut mirror_stream: TcpStream,
+        remote_rx: Receiver<Vec<u8>>,
+    ) {
+        let mut remote_stream = ReceiverStream::new(remote_rx);
+        let mut buffer = vec![0; 1024];
+
+        loop {
+            select! {
+                biased; // To allow local socket to be read before being closed
+
+                // Reads data that the user is sending from their socket to mirrord's interceptor
+                // socket.
+                read = mirror_stream.read(&mut buffer) => {
+                    match read {
+                        Err(fail) if fail.kind() == std::io::ErrorKind::WouldBlock => {
+                            continue;
+                        },
+                        Err(fail) => {
+                            error!("Failed reading mirror_stream with {:#?}", fail);
+                            break;
+                        }
+                        Ok(read_amount) if read_amount == 0 => {
+                            warn!("interceptor_task -> exiting due to local stream closed!");
+                            break;
+                        },
+                        Ok(read_amount) => {
+                            // Sends the message that the user wrote to our interceptor socket to
+                            // be handled on the `agent`, where it'll be forwarded to the remote.
+                            let write = Write { connection_id, bytes: buffer[..read_amount].to_vec() };
+                            let outgoing_data = TcpOutgoing::Write(write);
+
+                            if let Err(fail) = send_hook_message(HookMessage::TcpOutgoing(outgoing_data)).await {
```

Comment:
Why is this hook message? We're running from the `mirrord-layer` `Runtime` as a task here (AFAIK) - no need to use "bad" synch mechanisms - just pass a sender to this task from the main thread and bridge it.

Reply:
(Hook messages should be used from hook context, not tokio..)

Reply:
Makes sense, changed it to separate channels.